src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs

```diff
@@ -0,0 +1,379 @@
+        {
+            AssertAddThrowsOnNullAndEmpty("stateOrProvinceName", builder => builder.AddStateOrProvinceName);
+        }
+
+        [Fact]
+        public static void AddDomainComponent_Success_EncodeToIA5String()
+        {
+            AssertBuilder(
+                "301A31183016060A0992268993F22C6401191608494E5445524E414C",
+                builder => builder.AddDomainComponent("INTERNAL"));
+        }
+
+        [Fact]
+        public static void AddDomainComponent_NullOrEmpty_Fails()
+        {
+            AssertAddThrowsOnNullAndEmpty("domainComponent", builder => builder.AddDomainComponent);
+        }
+
+        [Fact]
+        public static void AddDomainComponent_InvalidIA5_Fails()
+        {
+            X500DistinguishedNameBuilder builder = new();
+            ArgumentException ex = AssertExtensions.Throws<ArgumentException>(
+                "domainComponent",
+                () => builder.AddDomainComponent("\u043Aevin"));
+            Assert.Contains("'IA5String'", ex.Message);
+        }
+
+        [Fact]
+        public static void AddEncoded_StringOid_Success()
+        {
+            AssertBuilder(
+                "301C311A3018061369F487D9C7B5D0AEA2CCCBE8C8C7F2F6F2BE27130141",
+                builder => builder.AddEncoded(TestOid, new byte[] { 0x13, 0x01, 65 }));
+
+            AssertBuilder(
+                "301C311A3018061369F487D9C7B5D0AEA2CCCBE8C8C7F2F6F2BE27130141",
+                builder => builder.AddEncoded(TestOid, stackalloc byte[] { 0x13, 0x01, 65 }));
+        }
+
+        [Fact]
+        public static void AddEncoded_Oid_Success()
+        {
+            AssertBuilder(
+                "301C311A3018061369F487D9C7B5D0AEA2CCCBE8C8C7F2F6F2BE27130141",
+                builder => builder.AddEncoded(new Oid(TestOid, null), new byte[] { 0x13, 0x01, 65 }));
+
+            AssertBuilder(
+                "301C311A3018061369F487D9C7B5D0AEA2CCCBE8C8C7F2F6F2BE27130141",
+                builder => builder.AddEncoded(new Oid(TestOid, null), stackalloc byte[] { 0x13, 0x01, 65 }));
+        }
+
+        [Fact]
+        public static void AddEncoded_InvalidDer()
+        {
+            X500DistinguishedNameBuilder builder = new();
+            AssertExtensions.Throws<ArgumentException>(
+                "encodedValue",
+                () => builder.AddEncoded(TestOid, new byte[] { 0xFF, 0xFF, 0xFF }));
+            AssertExtensions.Throws<ArgumentException>(
+                "encodedValue",
+                () => builder.AddEncoded(TestOid, stackalloc byte[] { 0xFF, 0xFF, 0xFF }));
+            AssertExtensions.Throws<ArgumentException>(
+                "encodedValue",
+                () => builder.AddEncoded(new Oid(TestOid), new byte[] { 0xFF, 0xFF, 0xFF }));
+            AssertExtensions.Throws<ArgumentException>(
+                "encodedValue",
+                () => builder.AddEncoded(new Oid(TestOid), stackalloc byte[] { 0xFF, 0xFF, 0xFF }));
+        }
```

Comment:
> 01 01 01 (throw, not DER-canonical)

`AsnDecoder` won't throw for this case. This goes back to my question in https://github.com/dotnet/runtime/issues/45169, which is basically that `AsnDecoder` doesn't really validate the contents of UNIVERSAL types.

I generally think this is okay, since `AddEncoded` is an escape hatch for "I know what I am doing". If we feel differently about that, I would be more interested in how we can improve the System.Formats.Asn1 functionality to do more DER validation and this can leverage that if/when that capability is added.

Reply:
Though, I would probably point out that if we don't restrict it now, it would be hard to further restrict it later. However, I still think the escape-hatch-esque utility for this still applies.

Reply:
Okay - wait. The ASN.1 APIs are still a little new to me. There is validation... I just need to actually move though it.

Reply:
I looked further in to this and... I generally feel that this is non-trivial and that something is missing from AsnReader, AsnDecoder, or something to further validate ASN.1 is actually DER by reading the whole blob. I went down the path of trying to validate the whole document here:

https://gist.github.com/vcsjones/4892cfeca11bfe9fb135772afdf9f35e

and started wondering if I am missing something with System.Formats.Asn1?

Reply:
The Remarks section of TryReadEncodedValue does say that it doesn't validate a lot (mainly that DER can't use indefinite encodings, etc).  https://docs.microsoft.com/en-us/dotnet/api/system.formats.asn1.asndecoder.tryreadencodedvalue?view=net-6.0#remarks

Perhaps we want an AsnDecoder::TryValidateEncodedValue, which would be the recursive thing you started writing there (and I think we have somewhere in CertificateRequest)